Contributing.md

dr-orlovsky opened the pull request:
We've being discussing that a `CONTRIBUTING.md` guidelines are needed for some time and I took an effort to propose it. The main text is taken from `rust-lightning` version (with some additions), but it is extended on the topics we were discussing in multiple issues previously:
- list of repository maintainers
- standard derives - #555
- naming conventions (after Bitcoin Core)
- MSRV and 2018 update #510
- code formatting with `rustfmt` #172

Each of these issue-related amendments to the basic (modified) version of rust-lightning contributing guidelines are made with a separate commit.

Kixunil commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,278 @@
+   #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+   ```
+   must be placed before each struct, and then those of these traits, which can't
+   be auto-derived because of the member field restrictions should be removed.
````
This should also forbid traits that conflict with invariants or guarantees of the type. Examples:

* structs with `Drop` (well, this is caught by the compiler but it's not **member field** restriction)
* Types that don't have reflexive equality/ordering (no `Eq`, `Ord`)

Possibly also forbid impls with foot-gunny or non-obvious implementations.

Also forbid **deriving** `Debug` on secrets - implement it manually by censoring the secret.

dr-orlovsky replied:
Added two points below addressing this issues: https://github.com/rust-bitcoin/rust-bitcoin/pull/587/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R206-R215

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+
+Note that Rust-LNPBP is currently considered "pre-production" during this time,
+there is no special handling of security issues. Please simply open an issue on
+Github.
```
Is it? At least `electrs` depends on it and I consider it in-production.

TheBlueMatt replied:
This appears to have been copied from LNPBP and isn't right.

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+Fuzzing is heavily encouraged: feel free to add related material under `fuzz/`
+
+Mutation testing is planned; any contribution there would be warmly welcomed.
+
```
Maybe add a requirement to test `unsafe` code with sanitizers including Miri?

dr-orlovsky replied:
Do we allow `unsafe` at all here?

dr-orlovsky replied:
Added a dedicated section addressing `unsafe`: https://github.com/rust-bitcoin/rust-bitcoin/pull/587/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R262-R266

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+achievable).
+
+
+## Coding conventions
```
Perhaps mention that Rust-idiomatic style is preferred to "how it looks in Core" if everyone agrees.

dr-orlovsky replied:
Done below

TheBlueMatt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+
+1. All non-error types should opportunistically derive, where it is possible,
+   the following traits:
+   - `Copy`
```
Maybe also mention this is only reasonable for "small" structs (in addition to the no-heap requirement).

Kixunil replied:
I'm not really sure if it helps in practice. Moves are `memcpy` anyway.

TheBlueMatt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+
+### MSRV
+
+The Minimal Supported Rust Version (MSRV) is 0.29; it is enforced by our CI. 
```
1.29 :)

dr-orlovsky replied:
Yeah, it feels like it is 0.29 :D

TheBlueMatt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,278 @@
+
+## Security
+
+Security is the primary focus of Rust-LNPBP; disclosure of security
```
Bitcoin :) and also below.

dr-orlovsky commented:
@TheBlueMatt @Kixunil addressed your comments with a separate commit 4e64db1e316ecd02d5d585f5561d90dd5544a981

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,297 @@
+   must be placed before each struct, and then those of these traits, which can't
+   be auto-derived because of the member field restrictions should be removed.
+
+2. `Eq`, `PartialEq`, `Ord`, `PartialOrd` derivation must be skipped/removed
+   from pt. 1 in the following situations:
+   - for types that don't have reflexive equality/ordering
+   - types which has lexicographic ordering defined as a part of a standard must
+     provide manual derivation
+   - types which may be more efficiently compared with bitcoin-specific rules
+     should provide manual derivation
+
+3. `Debug` must not be derived on structs and enums which may contain secret
+   data, and a manual `Debug` implementation should be provided instead.
```
I think "... implementation excluding private data should be ..." would be clearer

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,297 @@
+   from pt. 1 in the following situations:
+   - for types that don't have reflexive equality/ordering
+   - types which has lexicographic ordering defined as a part of a standard must
+     provide manual derivation
```
You mean manual *implementation*?

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,297 @@
+Security is the primary focus for this library; disclosure of security
+vulnerabilities helps prevent user loss of funds. If you believe a vulnerability
+may affect other  implementations, please inform them.
+
```
I think vulnerabilities in the library should be disclosed to the library authors as well. Contact information (with PGP key and cross-references to keybase/whatever) should be provided.

sgeisler replied:
Good question, currently we don't have a process for that I think. It seems like something that warrants some discussion. Are you proposing putting a list of email addresses+GPG keys into this document?

Kixunil replied:
I believe the standard is to put it into `SECURITY.md`, but at least linking it here would be useful.

sgeisler reviewed (commented):
Left some minor nits, looks good overall. Thx for tackling this issue :)

The only major thing we should talk about more are security disclosures imo.

dr-orlovsky commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,297 @@
+[Rust style guide][style guide] that has been formalized through the [style RFC
+process][fmt rfcs]. It is also required to run `cargo fmt` to make the code
+formatted according to `rustfmt` parameters 
+-->
```
Since it will be valid once we will reformat the crate with `rustfmt`, so after that this part can be just uncomented and not written from scratch

dr-orlovsky commented:
@Kixunil, @sgeisler updated with security suggestions

Kixunil commented:
ACK

sgeisler commented on Cargo.toml:
```diff
@@ -24,10 +24,10 @@ secp-recovery = ["secp256k1/recovery"]
 bech32 = "0.7.2"
 bitcoin_hashes = "0.9.1"
-secp256k1 = "0.20.0"
+secp256k1 = "0.20.1"
```
This seems very unrelated in this PR/commit.

dr-orlovsky replied:
Removed. It was not an oversight, but just me committing into the wrong branch (version update is required for error derives, which is parallel PR)

sgeisler reviewed (commented):
General ACK, just please revert the changes to the dependency versions (was probably just an oversight).

sgeisler reviewed (approved):
ACK 92bb0e503f34813ddb90bc4ad26a5a81d79571c4

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,301 @@
+Each commitment within a PR to the `master` must
+* compile without errors;
+* contain all necessary tests for the introduced functional;
+* contain all docs.
```
To be clear, each commit must compile and pass unit tests with no errors, with every feature combination (excepting `fuzztarget`, but including compiling the fuzztests) on some reasonably recent compiler. The tip of any PR branch must also compile and pass tests with no errors on 1.29.

I don't feel very strongly about insisting that commits include their own tests. I would mildly prefer that tests be in separate commits so that it's easy to reorder them during review and check that the new tests fail without the new code.

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,301 @@
+- [Sebastian Geisler](https://github.com/sgeisler)
+- [Sanket Kanjalkar](https://github.com/sanket1729)
+
+Anyone may become a prepository maintaner by:
```
Spelling, and also I'd rather drop this whole section. To become a maintainer I need to approve you and I have no explicit policy.

dr-orlovsky commented:
@apoelstra did all of that + tried to make CI scripts to match the guidelines as you described

sgeisler commented on contrib/test.sh:
```diff
@@ -2,6 +2,14 @@
+# Use toolchain if explicitly specified
+if [ -n "$TOOLCHAIN" ]
+then
+    alias cargo='cargo +"${TOOLCHAIN}"'
```
I wonder if we should replace `PIN_VERSIONS` with a check for the toolchain version? I removed that some time ago because the toolchain variable was used for nothing else, but now that might make sense again …

dr-orlovsky replied:
This will break the script once we update MSRV soon (and of course we will forget to update it here). Also script may be used for checking with custom toolchains and we cant do a reliable detection when we need to pin the versions

sgeisler replied:
> This will break the script once we update MSRV

It should, we should unpin as many deps as possible then. But yeah, when you don't want to set the toolchain that way it's still useful to have it as a separate flag.

dr-orlovsky replied:
I occasionally run tests on 1.36 w/o pins and it failed, so pinning probably will not change...

sgeisler commented on contrib/test.sh:
```diff
@@ -2,6 +2,14 @@
 FEATURES="base64 bitcoinconsensus use-serde rand"
 
+# Old cargo format is incompatible with a newer ones, so we need to clean this up
+rm -f Cargo.lock
```
Since this PR tells devs to use this script locally too I'm not sure if we should just delete the lockfile. Maybe print a warning instead if using an old rust version? The user will notice. Or at least gate it with a flag.

dr-orlovsky replied:
I will add NB about this to the CONTRIBUTING

dr-orlovsky replied:
Did that below

sgeisler commented on contrib/test.sh:
```diff
@@ -67,4 +70,7 @@ then
     cargo test --verbose
+
+    cd ..
+    rm -rf dep_test
```
Let's hope no one ever links `dep_test` to `/` :stuck_out_tongue_winking_eye:

dr-orlovsky replied:
If they do, the script will fail before, on `cargo new dep_test`

dr-orlovsky commented on contrib/ci.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh -ex
+
+git rebase -x 'AS_DEPENDENCY=true TOOLCHAIN=stable ./contrib/test.sh' master &&
```
You have to thank @apoelstra who was teaching in his reviews to do this constantly :)

sanket1729 reviewed (commented):
Would require an update to the IRC channels.

dr-orlovsky commented:
@sanket1729 did that

sanket1729 reviewed (commented):
One more small nit. The rest looks good.

dr-orlovsky commented:
@sanket1729 the comment does not show up, are you sure you have posted it?

dr-orlovsky commented:
Rebased onto the recent master. Seems like we already had very detailed discussion on this PR and it also provides improvements in CI standardization (test scripts etc), so I ask @apoelstra to merge it if possible

sanket1729 reviewed (commented):
ACK modulo squashing commits.

sanket1729 reviewed (commented):
ACK e0bdaf808ae861a508af6413f097ee67927ed223. Looks good, fixups should be squashed.

dr-orlovsky commented:
@sanket1729 done

apoelstra commented:
Almost every line of every commit has trailing spaces.

dr-orlovsky commented:
@apoelstra not sure I am getting it. Are you talking about commit messages or the .md file? Since I do not see spaces at the end of commit lines. If this is about .md file, do I need to re-do all the history and remove trailing spaces, is this critical?

sanket1729 commented:
ACk f198131. Reviewed that the last commit is only removing white spaces. Did not check whether it removed all of them :P

tobin-crypto commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+be reviewed.
+
+The current plan is to phase it in over multiple commits or even multiple PRs,
+which will introduce no other changes that re-formatting, such that each change
```
```suggestion
which will introduce no changes other than re-formatting, such that each change
```

tobin-crypto commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+
+## Going further
+
+You may be interested by Jon Atack guide on
```
```suggestion
You may be interested in the guide by Jon Atack on
```

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+   have default constructor initializing "empty" data structure, i.e. this
+   empty structure has a real use in the business logic *outside of the scope
+   of testing or creating dumb data*. For instance, if the structure consists
+   only of collection types which may be empty it should derive `Default` trait.
```
Randomly noticed I forgot about not implementing `Copy` on iterators.

dr-orlovsky replied:
Fixed

tcharding commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+requirements above is just a preliminary, but not necessary sufficient step for
+getting the PR accepted as a valid candidate PR for the `master` branch.
+
+PR authors may also find useful to run the following script locally in order to
+check that each of the commits within the PR satisfy the requirements above,
+before submitting the PR to review:
```
```suggestion
PR authors may also find it useful to run the following script locally in order
to check that each of the commits within the PR satisfies the requirements
above, before submitting the PR to review:
```

tcharding commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+- [Elichai Turkel](https://github.com/elichai)
+- [Sebastian Geisler](https://github.com/sgeisler)
+- [Sanket Kanjalkar](https://github.com/sanket1729)
+
+
```
```suggestion

```

tcharding commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+2. `Eq`, `PartialEq`, `Ord`, `PartialOrd` derivation must be skipped/removed
+   from pt. 1 in the following situations:
+   - for types that don't have reflexive equality/ordering
+   - types which has lexicographic ordering defined as a part of a standard must
+     provide manual implementation
+   - types which may be more efficiently compared with bitcoin-specific rules
+     should provide manual implementation
```
```suggestion
   - types which have a lexicographic ordering defined as a part of a standard
     must provide a manual implementation
   - types which may be more efficiently compared with bitcoin-specific rules
     should provide a manual implementation
```

tcharding commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+may affect other implementations, please disclose this information according to
+the [security guidelines](./SECURITY.md), work on which is currently in progress.
+Before it is completed, feel free to send disclosure to Andrew Poelstra,
+apoelstra@wpsoftware.net, and encrypted with his public key, which may be found
```
```suggestion
apoelstra@wpsoftware.net, encrypted with his public key, which may be found
```

tcharding commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,316 @@
+
+## Testing
+
+Related to the security aspect, rust bitcoin developers must be taking testing
+very seriously. Due to the modular nature of the project, writing new test cases
+is easy and good test coverage of the codebase is an important goal. Refactoring
+the project to enable fine-grained unit testing is also an ongoing effort.
```
```suggestion
Related to the security aspect, rust bitcoin developers take testing very
seriously. Due to the modular nature of the project, writing new test cases is
easy and good test coverage of the codebase is an important goal. Refactoring
the project to enable fine-grained unit testing is also an ongoing effort.
```

tcharding reviewed (commented):
Thanks @dr-orlovsky, I did a fairly grammatically picky review. My grammar is not excellent by any means so if you disagree with the suggestions feel free to ignore them. I have two additional nitpicks please
- Some times the file use two newlines between headings and sometimes a single line, can we pick one of these and be uniform across the file?
- Some lists use capitalization with ending period and some lists use no capitalization and no ending period. Can we pick one of these and be uniform also please?

Thanks

dr-orlovsky commented:
@Kixunil, @tcharding all your suggestions are added as last two commits - to siplify re-review for this large doc for those, who already reviewed and ACKed it before.

I propose to let it go the way it is, even if there are minor grammatical nits, since we can always do them as a separate PR and it is better to have non-grammar-perfect CONTRIBUTING guidelines than not to have them at all.

@sgeisler, @sanket1729 asking to re-ACK

@apoelstra, @TheBlueMatt pls review, since without it we can't finalize & merge this PR

tcharding reviewed (commented):
Great, thanks.

Kixunil commented:
ACK

Kixunil reviewed (approved):
I've re-read this again and I believe it's solid and would be better to merge this sooner rather than later. My comment about error types should be addressed in followup PR. I've also found a few nits but I don't think they are worth our time.

tcharding commented:
Friendly bump, please ack so we can merge :)

In no particular order: @sgeisler, @sanket1729, @apoelstra, @TheBlueMatt, @elichai

apoelstra commented:
Please rebase to
* FIx the gazillion trailing spaces on almost every line
* Bring fixes back to the commits they affect (e.g. Libera vs Freenode, Rust-LNPBP vs rust-bitcoin)

I also really don't like the rustfmt section. It is not "planned" to use rustfmt, there is a tracking issue which highlights several serious problems with rustfmt which have not been fixed in many years, and which will clearly not be fixed because the compiler tools community is too insular.

I also don't like the Derivation section. It is hard to follow and mostly summarized with "please use common sense". I think reviewers should just check on this on a case-by-case basis. Consider all the discussion happening on PRs to add `PartialEq` to errors, for example.

We should absolutely drop "Anyone can become a maintainer if they <meet some technical goals>". There are serious culture fit requirements.

There are several paragraphs, such as 
```Overall, this library must reflect Bitcoin Core approach whenever possible.
However, since many of the things in Bitcoin Core are maintained due to 
historical reasons and may represent poor design, Rust-idiomatic style is 
preferred to "how it looks in Core" if everyone agrees.
```
which appear to say nothing at all, and should be drop. There is also overuse of passive language and general flow concerns, though I won't hold up the PR because of those.

dr-orlovsky commented:
@apoelstra: I tried to address your suggestions as much as possible.

First, I separated commits unrelated to CONTRIBUTION into a new #767 - and squashed the rest. Previously, this guidelines was a copy from rust-lightning guidelines with changes on top of them - and later fixups. Since the amount of changes is already large, there is no reason for the reviewers to follow that history and we may start with a final text discussed on the previous round.

The squashing also removed "Anyone can become a maintainer if they" part which was actually already removed before in one of fixup commits.

Next, on top of a single initial commit I have applied changes that you have requested:
- Removed derivation section
- Rewrote section about formatting
- Improved style, removing passive language wherever possible.

I  propose to get this thing as an initial version and improve over it in the future.

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,317 @@
+> results by using the resources below.
+
+We have a dedicated developer channel on IRC, #rust-bitcoin@freenode.net where
+you may get helpful advice if you have questions.
```
In c7967c87b1856ef933ea556ebf466d68c0fd9487:

Still mentions freenode

apoelstra commented on CONTRIBUTING.md:
````diff
@@ -131,8 +128,8 @@ above, before submitting the PR to review:
 ```
-Where value in `BITCOIN_MSRV=1.29.0` should be replaced with the current MSRV
-from [README.md].
+Please replace the value in `BITCOIN_MSRV=1.29.0` with the current MSRV from 
````
In 7ac1fc5f6473838988a55b9fc65d965d3079ff46:

More trailing spaces.

dr-orlovsky replied:
How do you see them?
For me they are usually all get removed by `rustfmt`, but here we do not use that - and my eyes are not as laser as yours to see them clearly :D :D

Kixunil replied:
`grep ' $' README.md` should do the trick. :)

dr-orlovsky replied:
It did, but my eyes does not run `grep` as an internal extension. I tried CRISPR for that, but grep is still missed from existing CRISPR repos for human genes :D

apoelstra replied:
I use [this tool as my git-diff extension](https://github.com/so-fancy/diff-so-fancy) which highlights trailing whitespace in red.

I am hesitant to recommend it because of the obvious security implications of replacing your diff tool, but it seems to have been working great for me (and has never shown me a diff inconsistent with the normal diff utility, in the times that I've checked).

tcharding replied:
`git diff` shows trailing whitespace. You might need `--color=always` if you pass the output to a pager. FTR I use `vimpager` and have the following shell fuction along with a shell alias for it.
```
git-log-patch () {
    local commit
    if (( $# < 1 ))
    then
        commit="HEAD~"
    else
        commit="$1"
    fi
    git log --color=always --patch --reverse "${commit}".. | vimpager
}
```

Can be used like: `git-log-patch @~~` to get diffs of the last two patches.

dr-orlovsky commented:
@apoelstra ready for re-review

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -77,25 +77,23 @@ To contribute a patch, the workflow is a as follows:
 of GitHub's "assignee" field. First check that no one is assigned and then
 comment suggesting that you're working on it. If someone is already assigned,
-don't hesitate to ask if the assigned party or previous commenters are still
+don't hesitate to ask if the assigned party or previous commentators are still
```
In d581c9b546fb5b049aea7ba27f24edd1c44f811a:

I think `commenters` was correct.

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -226,7 +221,8 @@ the project to enable fine-grained unit testing is also an ongoing effort.
 Fuzzing is heavily encouraged: feel free to add related material under `fuzz/`
 
-Mutation testing is planned; any contribution there would be warmly welcomed.
+Mutation testing is planned; any contributions helipng with thet are highly
```
In d581c9b546fb5b049aea7ba27f24edd1c44f811a:

Typo `helipng`

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,317 @@
+ensure git bisect and other automated tools function properly.
+
+When adding a new feature thought must be given to the long term technical debt.
+Every new features should be covered by unit tests.
```
In 78d1a827697a757ccccf632280787f8a9be5a534:

`features` should be `feature`

apoelstra reviewed (approved):
ACK e1c8e13cb89d2b3fb732d317262c763a09f922f8

dr-orlovsky commented:
@Kixunil pls re-ACK. I updated the file according to Andrew's review. We have removed derive section to be added later as a separate PR after a proper discussion.

Kixunil reviewed (dismissed):
ACK e1c8e13cb89d2b3fb732d317262c763a09f922f8

dr-orlovsky commented:
@apoelstra shell we wait for more ACKs here - or this can be merged already?

apoelstra commented:
Let's wait for one more ACK. It's not going to conflict with anything.

RCasatta commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,236 @@
+Communication about Rust Bitcoin happens primarily in
+[#bitcoin-rust](https://web.libera.chat/?channel=#bitcoin-rust) IRC chat on
+[Libera](https://libera.chat/) with the logs available at
+<http://gnusha.org/rust-bitcoin/>.
```
It looks the logs at http://gnusha.org/rust-bitcoin/ didn't switch to libera because they are not working since about June 2021

dr-orlovsky replied:
Yes. We have to contact Bryan Bishop to do that.

apoelstra replied:
Sure, I'll ping him.

apoelstra replied:
Actually I just checked and it looks liek https://gnusha.org/bitcoin-rust/ is up to date.

So actually, we should change our CONTRIBUTING file and our IRC /topic.

dr-orlovsky replied:
May I do that with a fixup commit not to make re-review the whole PR for others?

Kixunil replied:
I vote for fixup, reviewing this is a huge pain already.

apoelstra replied:
Yep for sure.

RCasatta replied:
> Actually I just checked and it looks liek https://gnusha.org/bitcoin-rust/ is up to date.

Is it correct that days are there, but they are empty (except from log-start/log-end) up to June 2021?

apoelstra replied:
There is very little activity on the channel, but the logs do not appear empty to me.

RCasatta replied:
Ah sorry I was watching http://gnusha.org/rust-bitcoin/ while you linked http://gnusha.org/bitcoin-rust/ (rust and bitcoin switched)

dr-orlovsky replied:
Done

dr-orlovsky commented:
@apoelstra @Kixunil @RCasatta @sanket1729 I fixed the reference to IRC logs. Pls re-ack

Kixunil reviewed (approved):
ACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b

apoelstra reviewed (approved):
ACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b

dr-orlovsky commented:
Sorry for disturbing, @RCasatta @sanket1729, but this needs just a review of a single fixup commit in 31c49832b70f29f9d49ab9bb94050b68cc855f1b since your last reviews of this PR. We don't want to merge without your ACKs since you did a lot of commenting here before.

RCasatta commented:
ACK 31c4983

sanket1729 reviewed (approved):
utACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b. Looks good

dr-orlovsky commented:
Seems like we have 4 ACKs now (@RCasatta did it in a wrong way not clicking GitHub button, but I checked with bitcoin merge tool, it still sees/uses it). @apoelstra wouldn't you mind about merging it?

RCasatta commented:
I noticed one should take care using the word ACK in phrases not meant to be for acking otherwise the tool will take it for the "ACKs for top commit:" section even if it's not intended.

apoelstra commented:
I think, as long as you don't mention the top commit in the same line, you should be okay?

RCasatta commented:
yes, if you don't mention the top commit you are fine, pretty rare false positive, so it happened this time :)
